streams/src/test/java/org/apache/kafka/streams/processor/TopologyBuilderTest.java

```diff
@@ -94,6 +101,59 @@ public void testSourceTopics() {
         builder.addSource("source-2", "topic-2");
         builder.addSource("source-3", "topic-3");
 
-        assertEquals(builder.sourceTopics().size(), 3);
+        assertEquals(3, builder.sourceTopics().size());
     }
+
+    @Test
+    public void testTopicGroups() {
+        final TopologyBuilder builder = new TopologyBuilder();
+
+        builder.addSource("source-1", "topic-1", "topic-1x");
+        builder.addSource("source-2", "topic-2");
+        builder.addSource("source-3", "topic-3");
+        builder.addSource("source-4", "topic-4");
+        builder.addSource("source-5", "topic-5");
+
+        builder.addProcessor("processor-1", new MockProcessorSupplier(), "source-1");
+
+        builder.addProcessor("processor-2", new MockProcessorSupplier(), "source-2", "processor-1");
+        builder.copartitionSources(list("source-1", "source-2"));
+
+        builder.addProcessor("processor-3", new MockProcessorSupplier(), "source-3", "source-4");
+
+        List<List<String>> topicGroups = sort(builder.topicGroups());
+
+        assertEquals(3, topicGroups.size());
+        assertEquals(list(list("topic-1", "topic-1x", "topic-2"), list("topic-3", "topic-4"), list("topic-5")), topicGroups);
+
+        List<List<String>> copartitionGroups = sort(builder.copartitionGroups());
+
+        assertEquals(list(list("topic-1", "topic-1x", "topic-2")), copartitionGroups);
+    }
+
+    private List<List<String>> sort(Collection<Set<String>> topicGroups) {
```

Comment:
This code assumes there is not any duplicate topic in the input. So, I don' think it is not generic enough.